src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.Fixer.cs

```diff
@@ -0,0 +1,98 @@
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public class PreferStreamAsyncMemoryOverloadsFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
+            ImmutableArray.Create(PreferStreamAsyncMemoryOverloads.RuleId);
+
+        public sealed override FixAllProvider GetFixAllProvider() =>
+            WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            Document doc = context.Document;
+            CancellationToken ct = context.CancellationToken;
+            SyntaxNode root = await doc.GetSyntaxRootAsync(ct).ConfigureAwait(false);
+            if (root.FindNode(context.Span) is SyntaxNode node)
+            {
+                SemanticModel model = await doc.GetSemanticModelAsync(ct).ConfigureAwait(false);
+                if (model.GetOperation(node, ct) is IInvocationOperation invocation)
+                {
+                    string methodName = invocation.TargetMethod.Name;
+
+                    string title;
+                    if (methodName == "ReadAsync")
+                    {
+                        title = MicrosoftNetCoreAnalyzersResources.PreferStreamReadAsyncMemoryOverloadsTitle;
+                    }
+                    else if (methodName == "WriteAsync")
+                    {
+                        title = MicrosoftNetCoreAnalyzersResources.PreferStreamWriteAsyncMemoryOverloadsTitle;
+                    }
+                    else
+                    {
+                        return;
+                    }
+
+                    context.RegisterCodeFix(
+                        CodeAction.Create(
+                            title: title,
+                            createChangedDocument: c => FixInvocation(doc, root, invocation, methodName),
+                            equivalenceKey: MicrosoftNetCoreAnalyzersResources.PreferStreamAsyncMemoryOverloadsMessage),
+                        context.Diagnostics);
+                }
+            }
+        }
+
+        private static Task<Document> FixInvocation(Document doc, SyntaxNode root, IInvocationOperation invocation, string methodName)
+        {
+            SyntaxGenerator generator = SyntaxGenerator.GetGenerator(doc);
+
+            // The stream object
+            SyntaxNode instanceNode = invocation.Instance.Syntax;
+
+            // Need the byte array object so we can invoke its AsMemory() method
+            SyntaxNode bufferInstanceNode = invocation.Arguments[0].Value.Syntax; // byte[] buffer
+
+            // These arguments are not modified, just moved inside AsMemory
+            SyntaxNode offsetNode = invocation.Arguments[1].Syntax; // int offset
```

Comment:
Can we add unit tests where `offset` and `count` are specified in reverse order via named arguments? I think your implementation will handle these correctly and also maintain the named arguments for AsMemory invocation, but would be good to validate with unit test.

Reply:
Sure, I can do that. Good catch.

Reply:
@mavasani I added this test and it fails because the arguments get moved to the new `AsMemory` call with the exact same names that were being used in the ReadAsync/WriteAsync method, but the argument names are not the same in `AsMemory`:

ReadAsync/WriteAsync => int offset, int count
AsMemory => int start, int length

I think I should remove the name from the argument before adding it to the AsMemory call. I'm investigating how to do that. If you think I should do something differently, let me know, @mavasani.

Reply:
Can we just map the argument names, i.e. replace offset with start and count with length? Removing the name might not be desirable, especially if user prefers named arguments for constant values, which seem could be likely for these calls. You might have to add C# and VB specific code to check if argument syntax nodes have a [NameColon](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Generated/Syntax.xml.Syntax.Generated.cs,2059) and [this](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Editing/SyntaxGenerator.cs,1992) overload to create new argument syntax nodes.

Reply:
Basically, all arguments that had an explicit named argument should have a named argument in fixed code. All positional arguments without an explicit name should be retained as positional arguments.

Reply:
Understood. I'll make sure to keep the names. Thanks for the help.

Reply:
I also discovered that my fixer assumes the user will place the arguments in the correct location. If the cancellationToken or the buffer are located where the offset or the count are expected to be located, that's the argument the fixer ends up putting inside the AsMemory call. I'll address this.

Reply:
The analyzer expects ReadAsync/WriteAsync to have its arguments in this order:
- `byte[], int, int`
- `byte[], int, int, CancellationToken`

But if the user calls the method with named arguments in the wrong order (with the int data types in the wrong order as well), then the analyzer does not throw a diagnostic. Should I fix this or should we narrow down the analyzer to only work when the user puts the arguments in the correct order, or at the very least, the two `int` arguments in the 2nd and 3rd position?

Reply:
We should fix the analyzer. Use of named or positional arguments should not change the analysis behavior. I thought the IOperation API for invocation.Arguments abstracts out the user specified order in the sense that Arguments[0] always corresponds to the argument for the first parameter regardless of whether it was specified as the first argument by the user or at some other point using named argument. Do you not observe this?

Reply:
I looked at your analyzer implementation and I cannot figure out where in AnalyzeInvocation you perform any argument related checks- you only compare target method of invocation with one of the analyzed methods. That is correct implementation, and I would expect it to handle the case of out of order named arguments.

Reply:
> Do you not observe this?

I do not. I wrote a small unit test that attempts to fix the code example above and the arguments are not being found in the correct order, which makes me think I should find each argument by name, in the array of arguments, and I should not assume they will be located in the expected position.

Reply:
You are right about the Operation tree shape. `invocation.Arguments` are in source order, not in the parameter order: https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IArgument.cs#L145-L177

You would need to change your fixer to use `IArgumentOperation.Parameter.Ordinal` to identify the arguments corresponding to first two parameters (ordinal 0 and 1). You would need C# and VB specific implementations to identify if they are named arguments, and if so generate a named argument in fixed code. 

However, that still doesn't answer the question why your analyzer is not reporting a diagnostic for out of order named arguments case. It is not comparing any argument ordering, so I would expect the analyzer handle this case properly.

Reply:
> However, that still doesn't answer the question why your analyzer is not reporting a diagnostic for out of order named arguments case
You're correct. The analyzer does not have a problem, only the fixer. The unit test I wrote had the wrong `endColumn` specified, that was why the diagnostic was not calling the fixer.

Reply:
I fixed the named arguments case. Please take a look.